components/application-operator/pkg/kymahelm/gateway/gatewaymanager.go

```diff
@@ -0,0 +1,199 @@
+	return &gatewayManager{
+		helmClient:             helmClient,
+		overrides:              overrides,
+		serviceCatalogueClient: serviceCatalogueClient,
+		namespaces:             namespaceClient,
+	}
+}
+
+type gatewayManager struct {
+	helmClient             kymahelm.HelmClient
+	overrides              OverridesData
+	serviceCatalogueClient ServiceCatalogueClient
+	namespaces             v12.NamespaceInterface
+}
+
+func (g *gatewayManager) InstallGateway(namespace string) error {
+	overrides, err := kymahelm.ParseOverrides(g.overrides, overridesTemplate)
+	if err != nil {
+		return errors.Errorf("Error parsing overrides: %s", err.Error())
+	}
+
+	name := getGatewayName(namespace)
+
+	_, err = g.helmClient.InstallReleaseFromChart(gatewayChartDirectory, namespace, name, overrides)
+	if err != nil {
+		return errors.Errorf("Error installing Gateway: %s", err.Error())
+	}
+	return nil
+}
+
+func (g *gatewayManager) DeleteGateway(namespace string) error {
+	gateway := getGatewayName(namespace)
+	releaseExist, _, err := g.gatewayExists(gateway, namespace)
+	if err != nil {
+		return errors.Errorf("Error deleting Gateway: %s", err.Error())
+	}
+	if releaseExist {
+		return g.deleteGateway(gateway)
+	}
+	return nil
+}
+
+func (g *gatewayManager) deleteGateway(gateway string) error {
+	_, err := g.helmClient.DeleteRelease(gateway)
+	if err != nil {
+		return errors.Errorf("Error deleting Gateway: %s", err.Error())
+	}
+
+	return nil
+}
+
+func (g *gatewayManager) GatewayExists(namespace string) (bool, error) {
+	name := getGatewayName(namespace)
+	exists, _, err := g.gatewayExists(name, namespace)
+	return exists, err
+}
+
+func (g *gatewayManager) UpgradeGateways() error {
+	namespaces, err := g.getAllNamespacesWithServiceInstances()
+
+	if err != nil {
+		return errors.Errorf("Error updating Gateway: %s", err.Error())
+	}
+
+	if len(namespaces) == 0 {
+		return nil
+	}
+
+	err = g.updateGateways(namespaces)
+
+	if err != nil {
+		return errors.Errorf("Error updating Gateway: %s", err.Error())
+	}
+
+	return nil
+}
+
+func (g *gatewayManager) gatewayExists(name, namespace string) (bool, release.Status_Code, error) {
+	listResponse, err := g.helmClient.ListReleases(namespace)
+	if err != nil {
+		return false, release.Status_UNKNOWN, errors.Errorf("Error listing releases: %s", err.Error())
+	}
+
+	if listResponse == nil {
+		return false, release.Status_UNKNOWN, nil
+	}
+
+	for _, rel := range listResponse.Releases {
+		if rel.Name == name {
+			return true, rel.Info.Status.Code, nil
+		}
+	}
+	return false, release.Status_UNKNOWN, nil
+}
+
+func (g *gatewayManager) getAllNamespacesWithServiceInstances() ([]string, error) {
+	namespacesList, err := g.namespaces.List(metav1.ListOptions{})
+
+	if err != nil {
+		return nil, errors.Errorf("Error listing namespaces: %s", err.Error())
+	}
+
+	var namespaces []string
+
+	for _, namespace := range namespacesList.Items {
+		instances := g.serviceCatalogueClient.ServiceInstances(namespace.Name)
```

Comment:
I wonder if we can just pass `""` as a namespace to `ServiceInstances("")`. In core Kubernetes client we can then operate on resources from all namespaces. That would simplify this code.